### DISCUSSION TITLE
File-based external secrets

### DISCUSSION THREAD

**Discussion by ningw42** [navidrome/navidrome#3463] [Ideas]
We have `Spotify.Secret` and `LastFM.Secret` in configuration to specify secrets for external services. That makes the secrets available to anyone on the system that has read permission to the configuration file. One can set proper permission of the configuration file to block those access. But configuration file in general isn't supposed to readable just to the application.

The other common approach is to move secrets out of the main configuration file. By add something like `Spotify.SecretFile` to point to an external file with the secret's plaintext. Because secrets are intended to be read by the application only, we could set strict permission to protect it. This also comes with an extra benefits. If we someone configures system declaratively from a set of scripts, the external secrets can be managed separately, instead of being included in plaintext.

Here is my real-world use case. I am on [NixOS](https://nixos.org/), which manages system in a declarative way, from a set of `.nix` files. The NixOS community has [a list of common practices to manage secrets](https://nixos.wiki/wiki/Comparison_of_secret_managing_schemes), [sops-nix](https://github.com/Mic92/sops-nix) is one of them following the idea above leveraging [sops](https://github.com/getsops/sops). The general idea is to keep secrets encrypted at rest (e.g. a in VCS repo) but in plaintext when it is needed (i.e. a running system), so both the scripts and the secrets can be managed in VCS.

I took a look at the sources, there are two ways achieving this.
1. Read external secret files at configuration parsing time. Copy the secrets to its plaintext property. e.g `Spotify.Secret`. The rest are all the same. This is the simplest way, but this might not work for secrets that isn't a simple plaintext string. e.g. an encrypted file-based database in TOML/JSON/YAML needs further deserialization. But it's possible to just read the secret and delay the deserialization to agent.
2. Read external secret files at agent startup time. This delays the retrieval of secrets to each agent. And each agent can handle it differently. More flexible, but more work.

**Comment by deluan** [comment]
Not sure if it is worthy to add this complexity to Navidrome for these non-critical secrets. You can already get the secrets from wherever they are stored and pass them as env vars to ND, using a simple script wrapping the navidrome binary. Maybe this should be added to the Nix package?

**Reply by ningw42** [reply]
Thank you! That makes perfect sense. I was able to build a workaround with systemd's EnvironmentFile and Navidrome's environment variables support. 

I raised this because I see [Authelia has a built-in secret file support via environment variable](https://www.authelia.com/configuration/methods/secrets/#environment-variables). Given the different focus, I think it is reasonable for Authelia to have it natively but not for Navidrome.

**Reply by crazygolem** [reply]
This feature could also be useful for docker deployments: with [docker compose](https://docs.docker.com/compose/how-tos/use-secrets/) and [docker swarm](https://docs.docker.com/engine/swarm/secrets/#build-support-for-docker-secrets-into-your-images), secrets are provided as files. It has a few advantages over environment variables, e.g. allowing access control via file permissions (env vars set from docker are visible to all processes), less risk to being logged or leaked to another container inadvertently.

In this case too the workaround with a wrapper script would work.

**Reply by forksnd** [reply]
I too think this would be a good feature to have. It can be done the way MariaDB does it, where for any environment variable `ENV`, a variable `FILE__ENV` is tested and if it exists, it's value is interpreted as a filename and the content of that file (after stripping the trailing newline, if any) is treated as if it came from the `ENV` variable.
